pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -244,3 +450,120 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+func compareCAPIMachines(capiMachine1, capiMachine2 *capiv1beta1.Machine) []string {
+	var diff []string
+	diff = append(diff, deep.Equal(capiMachine1.Spec, capiMachine2.Spec)...)
+	diff = append(diff, util.ObjectMetaEqual(capiMachine1.ObjectMeta, capiMachine2.ObjectMeta)...)
+
+	return diff
+}
+
+// compareCAPIInfraMachines compares CAPI infra machines a and b, and returns a list of differences, or none if there are none.
+func compareCAPIInfraMachines(platform configv1.PlatformType, infraMachine1, infraMachine2 client.Object) ([]string, error) {
+	switch platform {
+	case configv1.AWSPlatformType:
+		typedInfraMachine1, ok := infraMachine1.(*capav1beta2.AWSMachine)
+		if !ok {
+			return nil, errAssertingCAPIAWSMachine
+		}
+
+		typedinfraMachine2, ok := infraMachine2.(*capav1beta2.AWSMachine)
+		if !ok {
+			return nil, errAssertingCAPIAWSMachine
+		}
+
+		var diff []string
+		diff = append(diff, deep.Equal(typedInfraMachine1.Spec, typedinfraMachine2.Spec)...)
+		diff = append(diff, util.ObjectMetaEqual(typedInfraMachine1.ObjectMeta, typedinfraMachine2.ObjectMeta)...)
+
+		return diff, nil
+	case configv1.PowerVSPlatformType:
+		typedInfraMachine1, ok := infraMachine1.(*capibmv1.IBMPowerVSMachine)
+		if !ok {
+			return nil, errAssertingCAPIIBMPowerVSMachine
+		}
+
+		typedinfraMachine2, ok := infraMachine2.(*capibmv1.IBMPowerVSMachine)
+		if !ok {
+			return nil, errAssertingCAPIIBMPowerVSMachine
+		}
+
+		var diff []string
+		diff = append(diff, deep.Equal(typedInfraMachine1.Spec, typedinfraMachine2.Spec)...)
+		diff = append(diff, util.ObjectMetaEqual(typedInfraMachine1.ObjectMeta, typedinfraMachine2.ObjectMeta)...)
+
+		return diff, nil
+	default:
+		return nil, fmt.Errorf("%w: %s", errPlatformNotSupported, platform)
+	}
+}
+
+// applySynchronizedConditionWithPatch updates the synchronized condition
+// using a server side apply patch. We do this to force ownership of the
+// 'Synchronized' condition and 'SynchronizedGeneration'.
+func (r *MachineSyncReconciler) applySynchronizedConditionWithPatch(ctx context.Context, mapiMachine *machinev1beta1.Machine, status corev1.ConditionStatus, reason, message string, generation *int64) error {
+	var severity machinev1beta1.ConditionSeverity
+	if status == corev1.ConditionTrue {
+		severity = machinev1beta1.ConditionSeverityNone
+	} else {
+		severity = machinev1beta1.ConditionSeverityError
+	}
+
+	conditionAc := machinev1applyconfigs.Condition().
+		WithType(consts.SynchronizedCondition).
+		WithStatus(status).
+		WithReason(reason).
+		WithMessage(message).
+		WithSeverity(severity)
+
+	util.SetLastTransitionTime(consts.SynchronizedCondition, mapiMachine.Status.Conditions, conditionAc)
+
+	statusAc := machinev1applyconfigs.MachineStatus().
+		WithConditions(conditionAc)
+
+	if status == corev1.ConditionTrue && generation != nil {
+		statusAc = statusAc.WithSynchronizedGeneration(*generation)
+	}
```

Comment:
If you send an apply and don't send the synchronized generation, it will clear the previously set synchronized generation, which I don't think is what we want, we want to copy over the old value when the condition is false I think.

Could you add a test for this?

It's probably worth having some unit testing for this function specifically so that we can be certain of it's behaviour